### ISSUE TITLE
Codierung von Sig-Voriger-Beleg

### ISSUE THREAD

**Issue by stabilokaefer** [a-sit-plus/at-registrierkassen-mustercode#376]
Wie in der Überschrift kurz erwähnt vermute ich in meiner Software ein Problem mit der Erstellung des Wertes Sig-Voriger-Beleg. Mein Problem ist es dass ich selbst kein Programmierer bin, sondern die Programmierung unserer Firma in Bulgarien ist und ich die Daten bestmöglich aufbereiten muss, dass ist mir denke ich soweit auch relativ gut gelungen.

Im Detailfragen Dok ist beschrieben, dass aus der Kompakte Repräsentation der JWS-Signatur der Hash Wert erstellt werden muss und die ersten 8 Bytes Base64 kodiert werden müssen.

Beim Startbelege wird die Kassen-ID dafür verwendet, in diesem Fall konnte ich den Wert wie im Dokument unter 2.4.1 Verkettung beim Startbeleg beschrieben nachstellen

Beispiel für KassenID: A12347

Hash Wert (http://hashgenerator.de/):
39e48a4233b8cca2c4161f9e0384dd442b97472eb790ad670b391a0da5aecd79

davon die ersten 8 Bytes:
39e48a4233b8cca2

diese base64 codiert (http://tomeko.net/online_tools/hex_to_base64.php?lang=en) :
OeSKQjO4zKI=

wenn ich nun das selbe für die Kompakte Repräsentation der JWS-Signatur versuche erhalte ich nicht den Wert wie in der Doku beschrieben

Beispiel:
eyJhbGciOiJFUzI1NiJ9.X1IxLUFUMF9ERU1PLUNBU0gtQk9YNTI0XzM2NjU4N18yMDE1LTEyLTE3VDExOjIzOjQ0XzM0LDc3XzU5LDY0XzM4LDEzXzAsMDBfMCwwMF84TUc4QzFLcjdIQT1fMjBmMmVkMTcyZGFhMDllNV94VGZadmtCU1RyND0.GeWps9kci-fUqKLymS1pHlIbv0L8Oek-v6TDmZj9Ffucb8yvSijqZ8LcBalV9lADMX Q8U3itViKkd_i1Ba22BA

Hash Wert:
2767baf8877a3dd1437648405fa666c1cd76bbc000e21005f15108089c14f702

davon die ersten 8 Bytes:
2767baf8877a3dd1

diese base64 codiert:
J2e6+Id6PdE=

In der Detailfragen Dokumenation wird der Wert 5HjRCx+XIz4= berechnet.

Das selbe verhalten konnte ich auch bei den Testcases des Prüftools nachstellen. hier konnte ich den Wert Sig-Voriger-Beleg für den StartBeleg nachvollziehen, allerdings nicht für die weiteren Belege

Ich hoffe ihr könnt mir weiterhelfen und verzeiht mein technisch eingeschänktes knowhow

**Comment by ErichFreitag** [comment]
Vielleicht hilft das:

![data process for signature creation](https://cloud.githubusercontent.com/assets/18228048/23518673/895748ca-ff74-11e6-8d9d-222ff6bc3a9b.jpg)

**Comment by ErichFreitag** [comment]
Übrigens steht da auch noch _R1-AT0 drinnen - das ist kein zulässiger VDA-Kennzeichner.

**Comment by stabilokaefer** [comment]
Hallo Erich,

danke für die Info. ich bin die Punkte von 4.5 über 2.4.2 zu 2.3 durchgegangen und habe auch die verknüpfungen berücksichtigt, konnte nun in meinem Vorgehen aber kein Problem finden.

Danke auch für den Hinweis dass der VDA Kinnzeichner nicht korrekt ist, dies ist mir schon aufgefallen. Wie gesagt beziehe ich mich auf das Beispiel vom Detailfragen Doc und da ist dieser VDA angegegeben.

vll können Sie mir anhand dieses Beispiels helfen, welches ich mit meinem Vorgehen auch nicht nachstellen kann, die prüfung mittels TestTool war OK. hier habe ich mit dem DemoClient von Atrust eine Rechnung erzeugt und folgende Zeile aus dem DEP genommen:

eyJhbGciOiJFUzI1NiJ9.X1IxLUFUMV9kZW1va2Fzc2U0Ml8yXzIwMTctMDMtMDJUMTc6MDI6MzlfNjUsMDBfMCwwMF8wLDAwXzAsMDBfMCwwMF8xbVBFQ3duRWJIc0VkdmszSzlrSXlnPT1fMjE1OUI5RjZfVUk3M01XcVRkVTQ9.TpK9NxgYRm2bFaTAMMjxVGPz9OkjevQOKjWJdBNpCfEpYuTZbHQ8YsQ3h2gu9cC2aibPYZ4P5Nu013SeKiHEWw

Payload decoded ergibt:
_R1-AT1_demokasse42_2_2017-03-02T17:02:39_65,00_0,00_0,00_0,00_0,00_1mPECwnEbHsEdvk3K9kIyg==_2159B9F6_UI73MWqTdU4=

hier ist der signaturwert des vorherigen Belegs UI73MWqTdU4=

wenn ich dies das versuche und die kompakte Sig representation hashe (http://hashgenerator.de/) erhalte ich:
7cb35558d67e6beabd3be923f9b29ec8ffa72c178b6a723f7f4bc55de7667cb8

davon die ersten 8 bytes: 7cb35558d67e6bea

und diese zu base64 codiere (http://tomeko.net/online_tools/hex_to_base64.php?lang=en) erhalte ich:
fLNVWNZ+a+o=

**Comment by ErichFreitag** [comment]
Das Beispiel mit den Online-Tools kann ich schon nachvollziehen, das hilft aber nicht weiter. Können sie mal einige Belege erzeugen und den DEP-Export generieren und das Ergebnis dann posten?

**Comment by jayarm** [comment]
"In der Detailfragen Dokumenation wird der Wert 5HjRCx+XIz4= berechnet."
 :)))))

Aber erhalte ich auch:
1. Beispiel :   J2e6+D96PdE= oder J2e6+Id6PdFDdkhAX6ZmwQ==
2. Beispiel :  fLNVWNZ+a+o= oder fLNVWNZ+a+q9O+kj+bKeyA==

**Comment by stabilokaefer** [comment]
es scheint dass diese werte nicht korrekt sind. ich bekomme über das testtool einen fehler zurück, irgendetwas macht unsere Software beim codieren der Standardbelege falsch..

**Comment by ErichFreitag** [comment]
Da sind noch mehrere Dinge seltsam (ohne Gewähr), vielleicht hängen die zusammen:
Beleg 1 - Umsatz 0, Umsatzzähler 0, Verkettung auf Kassen-ID OK
Beleg 2 - Umsatz 63 EUR, Umsatzzähler 419 EUR, Verkettung auf Beleg 1 OK
Beleg 3 - Umsatz 96 EUR, Umsatzzähler 419 EUR, Verkettung fehlerhaft
Umsatzzähler auch bei Folgebelegen immer 419 EUR.

Vielleicht wird die Verkettung von einem anderen (richtigen) Umsatzzählerwert zum falschen Zeitpunkt errechnet?

**Comment by stabilokaefer** [comment]
viellen Dank für den Hinweis mit dem Umsatzzähler, ich kann mir gut vorstellen, dass hier beim Export des DEP nicht der dazugehörige Wert der einzelnen Rechnung verwendet wird, sondern der derzeitige Stand.

dass werde ich gleich an unsere entwicklung weitergeben.

bei der prüfung des QR Codes für Beleg 2 ist mir aufgefallen, dass sich der Wert des Umsatzzählers auch zu jenen aus dem DEP unterscheidet.

leider erhalte ich auch bei der prüfung des qr codes mit dem testtool den Fehler CRYPTO_SIGNATURE Fail

_R1-AT1_1_2_2017-03-02T14:28:00_63,00_0,00_0,00_0,00_0,00_DhP1RLA8RDE8mQ8ADNR+fA==_C309DD7_R8NMVuSgxK0=_OuYZzT8kh0ZCoEEChZGczGVINmXoMRRzNusymtahhh9FRI6FA3CEq6vXX1sNRzKpCthCUaOfU1F8sP/3tVnSIQ==

Fällt Ihnen vielleicht auch bei dieser Zeile ein Problem auf, worauf der Fehler beruhen könnte?

**Comment by ErichFreitag** [comment]
Nein, eine Unstimmigkeit kann ich so nicht herauslesen, außer dass es die selbe Ursache haben könnte, weil es bei derselben Belegnummer auftritt.

**Comment by stabilokaefer** [comment]
danke für die hinweise, ich habe nun auserhalb vom system noch einmal mit dem webservice von atrust die signatur erstellt, dann habe ich beim prüfen des qr und des dep ein PASS erhalten. Ich habe zwar immer noch den Fehler, dass der Turnover Wert nicht stimmt, aber das werde ich nächste woche mit unserer entwicklung genauer kontrollieren.

**Comment by Josma** [comment]
Also mir kommt die zertifikatseriennummer (C309DD7) zu kurz vor.
Eventuell liegt hier ein Problem vor.
